joomla/joomla-cms
# Make Invalid Token warnings more user friendly

The session token validation implementation in various controller methods currently gives an empty white screen with "Invalid token" text. This is total uninformative to a normal user.

This update implicitly redirects to the previous page if the token is not valid. To prevent implicit redirect pass `false` as the second argument to  ~~`JSession::validateToken()`~~ `$this->checkToken()` from the `JControllerLegacy` controller instance. 

To make use of this method we need to replace any occurrence of  `JSession::checkToken('post') or jexit(JText::_('JINVALID_TOKEN'));` with ~~`JSession::validateToken('post');`~~ `$this->checkToken();`
## Summary of Changes
- Based on the discussion below I've removed the method from `JSession` class and put it in the closer context of its application viz. `JControllerLegacy`.
- All the controllers should now be able to simply call `$this->checkToken()` and the redirection shall be handled implicitly. **Implicit redirection is done to the referrer (calling) page always**.
- If a custom handling is required such as **alternate redirection url** or **custom response format like JSON** one may call `$this->checkToken` like:

``` php
if (!$this->checkToken('post', false))
{
     // ...
}
```

or even directly call the current `JSession::checkToken` (without the **die** part) like:

``` php
if (!JSession::checkToken('post'))
{
     // ...
}
```

 This will skip the implicit redirection and return false if the token does not match. Then the necessary `if - else` may be implemented as needed.
## Test Instruction

To see the effect the respective forms should be submitted after manipulating the token input in the form so as to trigger invalid token situation. This can usually be done by selecting in the browser "inspect element" (or similar) and modify the DOM.

If the token is valid (not manipulated or expired), everything should be working as usual without any difference.

The necessary changes are made in the front-end which should be tested:
- In `com_users` component try to register a new user, login, logout, remind username, reset password, save user profile.
- In `com_contact` submit the contact form to send an email.
- In `com_content` submit a vote.

---

**EDIT June 26, 2016:**
Added Summary of changes and Test Instructions based on the new changes.
